checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -1625,6 +1653,146 @@ && varTrackedInObligations(obligations, (LocalVariableNode) receiver))
     }
   }
 
+  /**
+   * Determines whether the given assignment is the first write to a private field during object
+   * construction, in order to suppress potential false positive resource leak warnings.
+   *
+   * <p>This method takes a conservative approach: it returns {@code true} only if it can
+   * definitively prove that this is the first assignment to the field during construction. Later
+   * assignments, if any, are analyzed independently and will raise errors if they overwrite
+   * unclosed resources. Therefore, we only check statements before this assignment.
+   *
+   * <p>The result is {@code true} if all the following hold:
+   *
+   * <ul>
+   *   <li>(1) The field is private.
+   *   <li>(2) It has no non-null inline initializer at its declaration.
+   *   <li>(3) It is not assigned in any instance initializer block.
+   *   <li>(4) The constructor does not use constructor chaining via {@code this(...)}.
+   *   <li>(5) There are no earlier assignments to the same field before this one in the
+   *       constructor.
+   *   <li>(6) There are no method calls before this assignment that might have already set the
+   *       field.
+   * </ul>
+   *
+   * @param field the field being assigned
+   * @param constructor the constructor where the assignment appears
+   * @param currentAssignment the actual assignment tree being analyzed, which is a statement in
+   *     {@code constructor}
+   * @return true if this assignment can be safely considered the first write during construction
+   */
+  private boolean isFirstAssignmentToField(
+      VariableElement field, MethodTree constructor, @FindDistinct Tree currentAssignment) {
+    TreePath constructorPath = cmAtf.getPath(constructor);
+    ClassTree classTree = TreePathUtil.enclosingClass(constructorPath);
+    String fieldName = field.getSimpleName().toString();
+
+    // (2) Disallow non-null inline initializer
+    for (Tree member : classTree.getMembers()) {
+      if (member instanceof VariableTree) {
+        VariableTree decl = (VariableTree) member;
+        if (decl.getName().contentEquals(fieldName) && decl.getInitializer() != null) {
+          if (decl.getInitializer().getKind() != Tree.Kind.NULL_LITERAL) {
+            return false;
+          }
+        }
+      }
+    }
+
+    // (3) Disallow assignment in instance initializer blocks
+    for (Tree member : classTree.getMembers()) {
```

Comment:
For efficiency, merge this into the above loop, like (5) and (6) are handled by a single loop.